File: src/zql/commands/get.zig
const std = @import("std");
const context = @import("../ctx.zig");
const types = @import("../../types.zig");
const collection = @import("../../collection.zig");
const mem = @import("memory");

pub fn get(ctx: *context.ZqlContext, allocator: std.mem.Allocator) anyerror!void {
    if (ctx.get_arg_count() != 1) {
        ctx.err = try std.fmt.allocPrint(allocator, "Expected 1 argument, got {}!", .{ctx.get_arg_count()});
        return;
    }

    var arg1 = ctx.get_arg(0).?;
    defer arg1.deinit();

    const user = ctx.get_user();
    var perm = try ctx.db.get_perm(user, "$", allocator);

    if (!perm.?.can_read) {
        ctx.err = try std.fmt.allocPrint(allocator, "Permissions denied, unable to read!", .{});
        return;
    }

    if (std.mem.eql(u8, arg1.value.String[0..1], "$")) {
        var v = try mem.AutoPtr(types.Value).init(allocator, .{ .Collection = ctx.db.* });
        defer v.deinit();
        v.set_on_free(types.dispose);
        ctx.buffer = v.move();
        return;
    }

    var split = std.mem.splitScalar(u8, arg1.value.String, '.');
    var current = ctx.db;

    while (split.next()) |s| {
        if (split.peek() == null) {
            var tmp_perm = try current.get_perm(user, s, allocator);

            if (tmp_perm) |p| {
                perm = p;
            }

            if (!perm.?.can_read) {
                ctx.err = try std.fmt.allocPrint(allocator, "Permissions denied, unable to read!", .{});
                return;
            }

            var val = current.get(s);

            if (val) |v| {
                defer v.deinit();
                ctx.buffer = v.move();
            } else {
                ctx.err = try std.fmt.allocPrint(allocator, "No such key \"{s}\"!", .{arg1.value.String});
            }
            break;
        }

        if (current.get(s)) |cur| {
            var tmp_perm = try current.get_perm(user, s, allocator);

            if (tmp_perm) |p| {
                perm = p;
            }

            if (!perm.?.can_read) {
                ctx.err = try std.fmt.allocPrint(allocator, "Permissions denied, unable to read!", .{});
                return;
            }

            switch (cur.value) {
                types.Value.Collection => {},
                else => {
                    ctx.err = try std.fmt.allocPrint(allocator, "Key \"{s}\" was expected to be Collection, but got {s}!", .{s, @tagName(cur.value)});
                    return;
                }
            }

            current = &cur.value.Collection;
            continue;
        }
        ctx.err = try std.fmt.allocPrint(allocator, "No such key \"{s}\"!", .{arg1.value.String});
        break;
    }
}

File: src/zql/commands/get.cpp
#include <cstdio>

#include <types.hpp>
#include <zql/ctx.hpp>

using ZeonDB::Types::Type;

void get_casual(ZeonDB::ZQL::Context* ctx) {
	auto key = ctx->get_arg(0);

	std::string user = ctx->get_user();
	auto perms = ctx->get_perm("$");

    auto* acc = ctx->get_account(user);
    if (acc && acc->special) {
        ctx->error = "Dummy can only create account!";
        return;
    }

	if (!perms.can_read) {
		ctx->error = "Permissions deniend, unable to read!";
		return;
	}

	if (key->v.s.compare("$") == 0) {
		*ctx->temporary_buffer = ctx->get_db();
		return;
	}

	auto current = ctx->get_db();
	std::string s = "";
	while ((s = key->v.s.next(".")).length() > 0) {
		if (!key->v.s.peek(".")) {
			if (current->v.c.has_perms(user, s)) {
				perms = current->v.c.get_perms(user, s);
			}

			if (!perms.can_read) {
				ctx->error = "Permissions deniend, unable to read!";
				return;
			}

			auto val = current->v.c.get(s);

			if (val == nullptr) {
				ctx->error = "No such key \"" + key->v.s + "\"!";
				return;
			}

			*ctx->temporary_buffer = val;
			break;
		}

		auto val = current->v.c.get(s);
		if (val != nullptr) {
			if (current->v.c.has_perms(user, s)) {
				perms = current->v.c.get_perms(user, s);
			}

			if (!perms.can_read) {
				ctx->error = "Permissions deniend, unable to read!";
				return;
			}

			if (val->t != Type::Collection) {
				ctx->error = "Collection expected at key " + s;
				return;
			}

			current = val;
			continue;
		}

		ctx->error = "No such key \"" + key->v.s + "\"!";
		break;
	}
}

void get(ZeonDB::ZQL::Context* ctx) {
	size_t arg_count = ctx->arg_count();
	switch (arg_count) {
		case 1:
			get_casual(ctx);
			break;
		default:
			ctx->error = R"(help: get (local) <key>
get <key> -- get value of <key> as local buffer
get local <key> -- get value of <key> from local buffer as local buffer

key syntax: key@branch[index]
- the ..range is optional)";
	}
}
